src/hotspot/share/jfr/support/methodtracer/jfrMethodTracer.cpp

```diff
@@ -0,0 +1,405 @@
+  assert(ik != nullptr, "invariant");
+  assert(!ik->is_scratch_class(), "invarint");
+  assert(ik->has_been_redefined(), "invariant");
+  assert(JfrTraceId::has_sticky_bit(ik), "invariant");
+  assert(in_use(), "invariant");
+  assert(SafepointSynchronize::is_at_safepoint(), "invariant");
+  assert_locked_or_safepoint(ClassLoaderDataGraph_lock);
+
+  const traceid klass_id = JfrTraceId::load_raw(ik);
+  const JfrInstrumentedClass jic(klass_id, ik, false);
+
+  if (instrumented_classes()->find(jic) == -1) { // not already existing
+    const int idx = instrumented_classes()->append(jic);
+    if (idx == 0) {
+      assert(!JfrTraceIdEpoch::has_method_tracer_changed_tag_state(), "invariant");
+      JfrTraceIdEpoch::set_method_tracer_tag_state();
+    }
+    add_timing_entry(ik, klass_id);
+    JfrTraceTagging::set_dynamic_tag_for_sticky_bit(ik);
+    log_add(ik);
+  }
+}
+
+#ifdef ASSERT
+static bool in_instrumented_list(const InstanceKlass* ik, const GrowableArray<JfrInstrumentedClass>* list) {
+  assert(ik != nullptr, "invariant");
+  assert(list != nullptr, "invariant");
+  assert_locked_or_safepoint(ClassLoaderDataGraph_lock);
+  const JfrInstrumentedClass jic(JfrTraceId::load_raw(ik), ik, false);
+  return list->find(jic) != -1;
+}
+#endif
+
+void JfrMethodTracer::add_instrumented_class(InstanceKlass* ik, GrowableArray<JfrTracedMethod>* methods) {
+  assert(ik != nullptr, "invariant");
+  assert(!ik->is_scratch_class(), "invariant");
+  assert(methods->is_nonempty(), "invariant");
+  ik->module()->add_read(jdk_jfr_module());
+  MutexLocker lock(ClassLoaderDataGraph_lock);
+  assert(!in_instrumented_list(ik, instrumented_classes()), "invariant");
+  JfrTraceTagging::set_dynamic_tag(ik, methods);
+  JfrTraceTagging::set_sticky_bit(ik, methods);
+  const JfrInstrumentedClass jik(JfrTraceId::load_raw(ik), ik, false);
+  const int idx = instrumented_classes()->append(jik);
+  if (idx == 0) {
+    JfrTraceIdEpoch::set_method_tracer_tag_state();
+  }
+  assert(in_instrumented_list(ik, instrumented_classes()), "invariant");
+  log_add(ik);
+}
+
+ModuleEntry* JfrMethodTracer::jdk_jfr_module() {
+  if (_jdk_jfr_module == nullptr) {
+    MutexLocker ml(Module_lock);
+    ModuleEntryTable* const table = Modules::get_module_entry_table(Handle());
+    Symbol* jfr_module_name = SymbolTable::probe("jdk.jfr", 7);
+    assert(jfr_module_name != nullptr, "jdk.jfr name could not be found");
+    _jdk_jfr_module = table->lookup_only(jfr_module_name);
+    assert(_jdk_jfr_module != nullptr, "jdk.jfr module could not be found");
+  }
+  return _jdk_jfr_module;
+}
+
+// Track the set of unloaded class ids during a chunk / epoch.
+static GrowableArray<jlong>* _unloaded_class_ids_0 = nullptr;
+static GrowableArray<jlong>* _unloaded_class_ids_1 = nullptr;
+static GrowableArray<jlong>* _current_unloaded_class_ids = nullptr;
+static GrowableArray<jlong>* _stale_class_ids = nullptr;
+static GrowableArray<jlong>* _empty_class_ids = nullptr;
```

Comment:
Global symbols, move them under `JfrMethodTracer`?

Reply:
static so not needed.